test/integration/etcd/restart_server.go

```diff
@@ -0,0 +1,489 @@
+		t.Fatal(err)
+	}
+
+	stopCh := make(chan struct{})
+	errCh := make(chan error, 100)
+
+	cleanup := start(t, stopCh, errCh, kubeAPIServerOptions, nil, customFlags, saSigningKeyFile, certDir, rawClient)
+
+	// need to do this AFTER start() to get the port
+	address := fmt.Sprintf("127.0.0.1:%d", kubeAPIServerOptions.SecureServing.BindPort)
+
+	kubeClientConfig := &restclient.Config{
+		Host: address,
+		TLSClientConfig: restclient.TLSClientConfig{
+			ServerName: "kubernetes.default.svc",
+			CAData:     caData,
+			CertData:   clientCertData,
+			KeyData:    clientKeyData,
+		},
+		QPS:   99999,
+		Burst: 9999,
+	}
+
+	kubeClient := clientset.NewForConfigOrDie(kubeClientConfig)
+	klog.Infof("kubeClientConfig.Host: %v", kubeClientConfig.Host)
+
+	err = healthCheck(t, kubeClient, errCh)
+	if err != nil {
+		// need to clean up here instead of test.cleanup since apiServer has not been created yet
+		if rawClient != nil {
+			rawClient.Close()
+		}
+		cleanup()
+		return nil, fmt.Errorf("healthCheck failed with err: %v", err)
+	}
+
+	// create CRDs so we can make sure that custom resources do not get lost
+	CreateTestCRDs(t, apiextensionsclientset.NewForConfigOrDie(kubeClientConfig), false, GetCustomResourceDefinitionData()...)
+
+	return &APIServer{
+		Client:           kubeClient,
+		Dynamic:          dynamic.NewForConfigOrDie(kubeClientConfig),
+		Config:           kubeClientConfig,
+		KV:               kvClient,
+		Cleanup:          cleanup,
+		ServerOpts:       kubeAPIServerOptions,
+		EtcdClient:       rawClient,
+		StopCh:           stopCh,
+		ErrCh:            errCh,
+		SaSigningKeyFile: saSigningKeyFile,
+		CertDir:          certDir,
+	}, nil
+}
+
+func healthCheck(t *testing.T, kubeClient *clientset.Clientset, errCh chan error) error {
+	lastHealth := ""
+	attempt := 0
+	if err := wait.PollImmediate(time.Second, time.Minute, func() (done bool, err error) {
+		select {
+		case err := <-errCh:
+			return false, err
+		default:
+		}
+
+		// wait for the server to be healthy
+		result := kubeClient.RESTClient().Get().AbsPath("/healthz").Do(context.TODO())
+		content, _ := result.Raw()
+		lastHealth = string(content)
+		if errResult := result.Error(); errResult != nil {
+			attempt++
+			if attempt < 10 {
+				t.Log("waiting for server to be healthy")
+			} else {
+				t.Log(errResult)
+			}
+			return false, nil
+		}
+		var status int
+		result.StatusCode(&status)
+		return status == http.StatusOK, nil
+	}); err != nil {
+		t.Log(lastHealth)
+		return err
+	}
+	return nil
+}
+
+// Restart restarts an API server
+func Restart(t *testing.T, customFlags []string, oldServer *APIServer) error {
+	klog.Infof("Restart() customFlags: %v", customFlags)
+	close(oldServer.StopCh)
+	stopCh := make(chan struct{})
+	errCh := make(chan error, 100)
+
+	oldServer.Cleanup = nil
+	cleanup := start(t, stopCh, errCh, oldServer.ServerOpts, &oldServer.ServerOpts.Etcd.StorageConfig, customFlags, oldServer.SaSigningKeyFile, oldServer.CertDir, oldServer.EtcdClient)
```

Comment:
This looks racey, how do you know the old server has shut down before you try starting the new one?